src/configurePipeline/services/source/sourceRepoService.ts

```diff
@@ -0,0 +1,172 @@
+                        break;
+                    case sourceOptions[1]:
+                    case sourceOptions[2]:
+                    default:
+                        repoGitConfigPath = vscode.workspace.workspaceFolders[0].uri.fsPath;
+                }
+                this.gitReference = git(repoGitConfigPath);
+                return this.getGitRepoDetails(azureDevOps);
+            })
+            .then((gitRepoDetails) => {
+                return gitRepoDetails;
+            });
+    }
+
+    /**
+     *
+     * @param pipelineYamlPath : local path of yaml pipeline in the extension
+     * @param context: inputs required to be filled in the yaml pipelines
+     * @returns: thenable object which resolves once all files are added to the repository
+     */
+    public addYmlFileToRepo(ymlFilePath: string, context: any) {
+        let deferred: Q.Deferred<string> = Q.defer();
+        fs.readFile(ymlFilePath, { encoding: "utf8" }, async (error, data) => {
+
+            if (error) {
+                vscode.window.showErrorMessage(error.message);
+            }
+            else {
+                let fileContent = Mustache.render(data, context);
+                let ymlFileUri = vscode.Uri.file(vscode.workspace.workspaceFolders[0].uri.path + "/azure-pipelines.yml");
+                fs.writeFileSync(ymlFileUri.fsPath, fileContent);
+                await vscode.workspace.saveAll(true);
+                deferred.resolve(ymlFileUri.fsPath);
+            }
+        });
+
+        return deferred.promise;
+    }
+
+    /**
+     * commits yaml pipeline file into the local repo and pushes the commit to remote branch.
+     * @param pipelineYamlPath : local path of yaml pipeline in the repository
+     * @returns: thenable object which resolves once commit is pushed to remote branch, and failure message if unsuccessful
+     */
+    public async commitAndPushPipelineFile(pipelineYamlPath: string): Promise<{ commitId: string, branch: string }> {
+        // TODO
+        // unstages the changes already staged
+        // stages the file at the mentioned path
+        // commit the staged file
+        // stage the previously staged changes
+
+
+        await this.gitReference.add(pipelineYamlPath);
+        let commit = await this.gitReference.commit("Add yml file to workspace", pipelineYamlPath);
+        let status = await this.gitReference.status();
+        let remote = status.tracking;
+        let branch = status.current;
+        if (remote && branch) {
+            remote = remote.substr(0, remote.indexOf(branch) - 1);
+            await this.gitReference.push(remote, branch);
+        }
+        else {
+            throw new Error("Cannot add yml file to your git repository, remote is not set");
+        }
+        return {
+            branch: branch,
+            commitId: commit.commit
+        };
+    }
+
+    private async getGitRepoDetails(azureDevOps: AzureDevOpsService): Promise<GitRepoDetails> {
```

Comment:
IMO azureDevOpsService instance as a parameter does not make sense at this point as we are yet to derive organization and repo info. Its blank object that we are passing around to eventually call set method. Can we pass the credentials here and create instance of AzureDevOpsService once we have repo info and set extensionVariables.azureDevOpsService ? or maybe create instance later on when required.